File: ojuz/boi/boi-2017/boi2017_solutions/friends/Friends_old.java

import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.Stack;

public class Friends 
{
	public static int n; 
	public static int p;
	public static int q; 
	public static ArrayList<ArrayList<Integer>> graph; 
	
	public static ArrayList<Boolean> visited; 
	public static ArrayList<Boolean> in; 
	public static ArrayList<Boolean> undecided; 
	public static Stack<Integer> stack; 
	public static ArrayList<Boolean> out; 
	
	public static void main(String[] args) throws Exception
	{
		String line = null;
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		line = br.readLine();

		n = Integer.parseInt(line.split(" ")[0]);
		p = Integer.parseInt(line.split(" ")[1]);
		q = Integer.parseInt(line.split(" ")[2]);
		graph = new ArrayList<ArrayList<Integer>>();
		
		for(int i = 0; i < n; i++)
		{
			ArrayList<Integer> temp = new ArrayList<Integer>();
			line = br.readLine(); 
			String[] split = line.split(" ");
			for(int j = 1; j < split.length; j++)
				temp.add(Integer.parseInt(split[j]));
			graph.add(temp);
		}
		
		//System.out.println(n + " " + p + " " + q);
		//System.out.println(graph);
		long startTime = System.currentTimeMillis();
		
		visited = new ArrayList<Boolean>();
		in = new ArrayList<Boolean>();
		undecided = new ArrayList<Boolean>();
		out = new ArrayList<Boolean>();
		
		for(int i = 0; i < n; i++)
		{
			visited.add(false);
			in.add(false);
			undecided.add(false);
			out.add(false);
		}
		
		stack = new Stack<Integer>(); 
		boolean success = true;
		
		for(int u = 0; u < n && success; u++) 
		{
			for(int i = 0; i < graph.get(u).size(); i++) 
			{
				int v = graph.get(u).get(i);
				boolean foundu = false;
				for(int j = 0; j < graph.get(v).size(); j++) 
				{
					foundu |= (graph.get(v).get(j) == u);
				}
				success &= foundu;
			}
		}
		
		for(int i = 0; i < n; i++) 
		{
			if(!visited.get(i) && !findgroup(i)) 
			{
				success = false;
				break;
			}
		}
		
		if(success) 
			System.out.println("home"); 
		else 
			System.out.println("detention");
		
		long endTime = System.currentTimeMillis();
    // System.out.println("That took " + (endTime - startTime) / 1000.0 + " seconds");
	}
	
	public static int popUndecided()
	{
		int v = stack.pop();
		undecided.set(v, false);
		return v;		
	}

	public static void pushUndecided(int v) 
	{
		undecided.set(v, true);
		stack.push(v);
	}

	public static ArrayList<Integer> freshNeighbors(int v) 
	{
		ArrayList<Integer> ans = new ArrayList<Integer>();
		
		for(int i = 0; i < graph.get(v).size(); i++) 
		{
			int neighbor = graph.get(v).get(i);
			if(!(in.get(neighbor) || undecided.get(neighbor) || out.get(neighbor))) 
				ans.add(neighbor);
		}
		
		return ans;
	}
	
	public static boolean findgroup(int v) 
	{
		pushUndecided(v);
		boolean ans = branchIn(0,0);
		popUndecided();
		visited.set(v, true);
		return ans;
	}
	
	public static boolean mainBranch(int groupSize, int cutSize) 
	{
		if(groupSize > p || cutSize > q || groupSize + cutSize + stack.size() > p+q) 
			return false;
		if(stack.empty()) 
			return true;
		return(branchOut(groupSize, cutSize) || branchIn(groupSize, cutSize));
	}

	public static boolean branchOut(int groupSize, int cutSize) 
	{
		int v = popUndecided();
		out.set(v, true); 
		for(int i = 0; i < graph.get(v).size(); ++i) 
		{
			if(in.get(graph.get(v).get(i)))
				cutSize++;
		}

		boolean ans = mainBranch(groupSize, cutSize);
		out.set(v, false);
		pushUndecided(v);
		return ans;
	}

	public static boolean branchIn(int groupSize, int cutSize) 
	{
		int v = popUndecided();
		in.set(v, true);

		ArrayList<Integer> nv = freshNeighbors(v);
		for(int i = 0; i < nv.size(); ++i) 
			pushUndecided(nv.get(i));
		
		for(int i = 0; i < graph.get(v).size(); ++i) 
		{
			if(out.get(graph.get(v).get(i)))
				cutSize++;
		}
		
		boolean ans = mainBranch(groupSize + 1, cutSize);
		in.set(v, false);
		for(int i = 0; i < nv.size(); ++i) 
			popUndecided();
		pushUndecided(v);

		if(ans)	
			visited.set(v, true);
		return ans;
	}
}

File: ojuz/boi/boi-2017/boi2017_solutions/friends/accepted/friends_old.cpp
#include<stdio.h>
#include<vector>
#include<iostream>
#include<stack>

using namespace std;

int n,p,q;
vector<vector<int> > G;

vector<bool> visited; // already put in some previous group
vector<bool> in; // active group
vector<bool> undecided; // neighbors of active group, may go into active group.
stack<int> undecidedStack; // a stack of the yet undecided neighbors of the active group
vector<bool> out; // neighbors of active group, may NOT go into active group.

bool branchWithFirstVertexOut(int groupSize, int neighSize);
bool mainBranch(int groupSize, int neighSize);
bool branchWithFirstVertexInGroup(int groupSize, int neighSize);
bool findgroup(int v);

// ----------------------------------------
// ---------------------------------------------- MAIN GREEDY
// ----------------------------------------

int main() {

	scanf("%d %d %d",&n, &p, &q);
	G = vector<vector<int> > ();
	bool success = true;

	// read graph	
	for(int i = 0; i < n && success; ++i) {
		int mi;
		scanf("%d", &mi);
		if(mi > p + q) success = false;
		G.push_back(vector<int> (mi, 0));
		for(int j = 0; j < mi; ++j) {
			scanf("%d", &(G[i][j]));
		}
	}
	
	// verify undirected
	for(int u = 0; u < n && success; ++u) {
		for(int i = 0; i < G[u].size(); ++i) {
			int v = G[u][i];
			bool foundu = false;
			for(int j = 0; j < G[v].size(); ++j) {foundu |= (G[v][j] == u);}
			success &= foundu;
		}
	}
	
	// check that each vertex can be put in some group.
	visited = vector<bool> (n, false);
	in = vector<bool> (n, false);
	undecided = vector<bool> (n, false);
	out = vector<bool> (n, false);
	
	for(int i = 0; i < n && success; ++i) {
		if(!visited[i] && !findgroup(i)) {
			success = false;
			break;
	}}
	
	if(success) printf("home\n"); else printf("detention\n");
	
	return 0;
	
}

// ----------------------------
// ---------------------- HELPER FUCNTIONS FOR BRANCHING ALGORITHM BELOW
// ----------------------------

inline int popUndecided() {
	int v = undecidedStack.top();
	undecidedStack.pop();
	undecided[v] = false;
	return v;
}

inline void pushUndecided(int v) {
	undecided[v] = true;
	undecidedStack.push(v);
}

vector<int> freshNeighbors(int v) {
	vector<int> ans;
	for(int i = 0; i < G[v].size(); ++i) {
		if(!(in[G[v][i]] || undecided[G[v][i]] || out[G[v][i]])) ans.push_back(G[v][i]);
	}
	return ans;
}

// ------------------------------------------------- BRANCHING
// ------------------------------------------------- FINDING IF V CAN BE PUT IN A GROUP
bool findgroup(int v) {
	pushUndecided(v);
	bool ans = branchWithFirstVertexInGroup(0,0);
	popUndecided();
	visited[v] = true;
	return ans;
}

// groupsize is the current size of the group, cutsize is the size of the cut
bool mainBranch(int groupSize, int cutSize) {

	if(groupSize > p || cutSize > q || groupSize + cutSize + undecidedStack.size() > p+q) return false;
	if(undecidedStack.empty()) {return true;}
	return(branchWithFirstVertexOut(groupSize, cutSize) || branchWithFirstVertexInGroup(groupSize, cutSize));

}

bool branchWithFirstVertexOut(int groupSize, int cutSize) {
	int v = popUndecided();
	out[v] = true; 
	for(int i = 0; i < G[v].size(); ++i) {cutSize += in[G[v][i]];}

	bool ans = mainBranch(groupSize, cutSize);
	
	out[v] = false;
	pushUndecided(v);
	
	return ans;
}

bool branchWithFirstVertexInGroup(int groupSize, int cutSize) {
	int v = popUndecided();
	in[v] = true;

	vector<int> nv = freshNeighbors(v);
	for(int i = 0; i < nv.size(); ++i) {pushUndecided(nv[i]);}
	for(int i = 0; i < G[v].size(); ++i) {cutSize += out[G[v][i]];}
	
	bool ans = mainBranch(groupSize + 1, cutSize);
	
	in[v] = false;
	for(int i = 0; i < nv.size(); ++i) {popUndecided();}
	pushUndecided(v);

	if(ans)	visited[v] = true; // If found a group with v in it, v becomes visited. 
	
	return ans;
}
